urlfaa9045c2e_Azmisov_racko_ai_tgz-pJ8-client_TestingJ8/src/models/ModelAI.java

```diff
@@ -0,0 +1,366 @@
+package models;
+
+import NeuralNetworks.Network;
+import interfaces.Model;
+import interfaces.Player;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Random;
+import racko.Game;
+import racko.Rack;
+
+/**
+ * Plays the game as an artificial intelligence
+ */
+public class ModelAI extends Model{
+	private static final double LEARN_RATE = .1, EPSILON = 0.0001;
+	private static final Random RAND = new Random();
+	//Playing history
+	private final ArrayList<DataInstance> drawHistory;
+	private final ArrayList<DataInstance> playHistory;
+	private DataInstance draw_instance, play_instance;
+	//Learning model
+	private final boolean USE_RAND, USE_PROB_DRAW = false, USE_PROB_PLAY = false;
+	private final int RAND_LIMIT = 20, RAND_ROUNDS = 0;
+	public Network drawNet = null, playNet = null;
+	private final String drawNet_file, playNet_file;
+	private final boolean TRAIN;
+	//Deep learning
+	private final int DL_maxlayers = 4;
+	private int DL_drawdelta, DL_playdelta, DL_layers = 0, rack_size;
+	private final StoppingCriteria DL_stop = new StoppingCriteria(04, 50);
+	//Statistics
+	private double initialScore, currentScore;
+	private int games_played = 0, net_play_count, moves_in_round;
+	public int rand_count = 0;
+	
+	/**
+	 * Load AI from stored file
+	 * @param draw_file weights for draw network
+	 * @param play_file weights for play network
+	 * @param rack_size rack size
+	 * @param train should we train the network?
+	 * @param random play random moves to explore the feature space
+	 */
+	public ModelAI(String draw_file, String play_file, int rack_size, boolean train, boolean random){
+		USE_RAND = random;
+		this.rack_size = rack_size;
+		drawNet_file = draw_file;
+		playNet_file = play_file;
+		File draw_f = new File(drawNet_file),
+			play_f = new File(playNet_file);
+		boolean draw_loaded = false,
+				play_loaded = false;
+		if (draw_f.isFile()){
+			try{
+				if (drawNet.inputNodes() != calculateNodeCount(true, true, rack_size) ||
+					drawNet.outputNodes() != 1)
+					throw new Exception("Invalid input/output size");
+				draw_loaded = true;
+			} catch (Exception e){
+				System.out.println(e.getMessage());
+				System.out.println("Warning!!! Could not load AI draw network weights");
+			}
+		}
+		if (draw_f.isFile()){
+			try{
+				playNet = new Network(playNet_file);
+				if (playNet.inputNodes() != calculateNodeCount(true, false, rack_size) ||
+					playNet.outputNodes() != rack_size+1)
+					throw new Exception("Invalid input/output size");
+				play_loaded = true;
+			} catch (Exception e){
+				System.out.println(e.getMessage());
+				System.out.println("Warning!!! Could not load AI play network weights");
+			}
+		}
+		if (!draw_loaded)
+			newDrawNetwork();
+		if (!play_loaded)
+			newPlayNetwork();
+		//Training
+		TRAIN = train;
+		if (TRAIN){
+			drawHistory = new ArrayList();
+			playHistory = new ArrayList();
+			initDeepLearning();
+		}
+		else{
+			drawHistory = null;
+			playHistory = null;
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 83:91, 109:117.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.